casper/contracts/simple_casper.v.py

```diff
@@ -428,6 +445,50 @@ def highest_finalized_epoch(min_total_deposits: wei_value) -> int128:
     # to signal not to locally finalize anything
     return -1
 
+
+@public
+@constant
+def havent_voted_and_source_target_right(
```

Comment:
I see what you're doing here with `havent_voted_and_source_target_right`. We don't want to have to repeat parse all the vars in `vote` again in the helper function. I think there's a better way. We can essentially have a private version of votable -- `_votable` that takes the params rather than the vote_msg and both `votable` and `vote` call it.

What if we do something like
```python
@private
def _votable(validator_index, target_hash, target_epoch, source_epoch):
    check_not_voted()
    check_correct_target_epoch_and_hash()
    check_in_current_or_prev_dyn()

@public
def votable(vote_msg):
    vars = parse_vars
    return _votable(validator_index, target_hash,...)

def vote(vote_msg):
    vars = parse_vars
    assert _votable(validator_index, target_hash,...)
    assert validate_vote_msg(vote_msg)
    
    parse_any_other_needed_vars
    do_vote_logic()
```

There is a bit of repeat in the variables that need to be parsed, but not too much. I would rather get it readable now, and only trim it up in the future if we need to cut gas costs.